src/generator/extract_schema.cpp

```diff
@@ -0,0 +1,323 @@
+{
+    using underlying_type = typename std::underlying_type<scalar_type>::type;
+    auto value = std::hash<underlying_type>{}(static_cast<underlying_type>(node.type));
+    for (const auto &[k, v] : node.tags) {
+        value ^= std::hash<std::string_view>{}(k) ^ std::hash<std::string_view>{}(v);
+    }
+    return value;
+}
+
+std::size_t node_hash::operator()(const node_array_ptr &node) const noexcept
+{
+    std::size_t value = std::hash<bool>{}(node->truncated) ^ std::hash<std::size_t>{}(node->length);
+    for (const auto &child : node->children) { value ^= std::visit(node_hash{}, child); }
+    return value;
+}
+
+std::size_t node_hash::operator()(const node_record_ptr &node) const noexcept
+{
+    std::size_t value = std::hash<bool>{}(node->truncated);
+    for (const auto &[key, child] : node->children) {
+        value ^= std::hash<std::string_view>{}(key) ^ std::visit(node_hash{}, child);
+    }
+    return value;
+}
+
+bool node_equal::operator()(const node_scalar &lhs, const node_scalar &rhs) const
+{
+    return lhs.type == rhs.type && lhs.tags == rhs.tags;
+}
+
+bool node_equal::operator()(const node_array_ptr &lhs, const node_array_ptr &rhs) const
+{
+    if (lhs->length != rhs->length || lhs->children.size() != rhs->children.size() ||
+        lhs->truncated != rhs->truncated) {
+        return false;
+    }
+
+    // NOLINTNEXTLINE(readability-use-anyofallof)
+    for (const auto &node : lhs->children) {
+        if (rhs->children.find(node) == rhs->children.end()) {
+            return false;
+        }
+    }
+    return true;
+}
+
+bool node_equal::operator()(const node_record_ptr &lhs, const node_record_ptr &rhs) const
+{
+    if (lhs->children.size() != rhs->children.size() || lhs->truncated != rhs->truncated) {
+        return false;
+    }
+
+    auto lhs_it = lhs->children.begin();
+    auto lhs_end = lhs->children.end();
+
+    auto rhs_it = rhs->children.begin();
+    auto rhs_end = rhs->children.end();
+
+    for (; lhs_it != lhs_end && rhs_it != rhs_end; ++lhs_it, ++rhs_it) {
+        if (rhs_it->first != lhs_it->first) {
+            return false;
+        }
+        if (!std::visit(node_equal{}, lhs_it->second, rhs_it->second)) {
+            return false;
+        }
+    }
+
+    return true;
+}
+
+bool node_equal::operator()(const base_node &lhs, const base_node &rhs) const
+{
+    return std::visit(node_equal{}, lhs, rhs);
+}
+
+struct node_serialize {
+    ddwaf_object operator()(const std::monostate & /*node*/) const noexcept;
+    ddwaf_object operator()(const node_scalar &node) const noexcept;
+    ddwaf_object operator()(const node_array_ptr &node) const noexcept;
+    ddwaf_object operator()(const node_record_ptr &node) const noexcept;
+};
+
+ddwaf_object node_serialize::operator()(const std::monostate & /*node*/) const noexcept
+{
+    ddwaf_object tmp;
+    ddwaf_object array;
+    ddwaf_object_array(&array);
+    ddwaf_object_array_add(&array, ddwaf_object_unsigned(&tmp, 0));
```

Comment:
`0` is the type value for `unknown` in the spec, I didn't add it to the enum to avoid having two ways of defining an unknown type, i.e. through `std::monostate` and through a scalar of type unknown, although I can add a comment or a constexpr with a name that makes sense.